{
  "nbformat": 4,
  "nbformat_minor": 0,
  "metadata": {
    "colab": {
      "provenance": []
    },
    "kernelspec": {
      "name": "python3",
      "display_name": "Python 3"
    },
    "language_info": {
      "name": "python"
    }
  },
  "cells": [
    {
      "cell_type": "markdown",
      "source": [
        "# Introduction\n",
        "\n",
        "**Goal:** connect biology and coding immediately.\n",
        "\n",
        "__Short markdown cell__:\n",
        "\n",
        "“In this session, we’ll see how Python helps us analyze biological data — from DNA sequences to molecular properties — using simple programming tools.”\n",
        "\n",
        "__Example biological question__:\n",
        "\n",
        "- “How can we write a small program to count GC content in DNA sequences?”\n",
        "- “How can we store and look up gene information efficiently?”"
      ],
      "metadata": {
        "id": "qTE5hxGkQArj"
      }
    },
    {
      "cell_type": "markdown",
      "source": [
        "# Variables and Data Types"
      ],
      "metadata": {
        "id": "XwAbvubGUOgz"
      }
    },
    {
      "cell_type": "code",
      "execution_count": 8,
      "metadata": {
        "colab": {
          "base_uri": "https://localhost:8080/"
        },
        "id": "KlS26GqBPRH7",
        "outputId": "b12cca6a-741f-4fa1-ed81-265b6db83199"
      },
      "outputs": [
        {
          "output_type": "stream",
          "name": "stdout",
          "text": [
            "sample_name: E_coli_geneA\n",
            "type(sample_name): <class 'str'>\n",
            "\n",
            "dna_length: 5000\n",
            "type(dna_length): <class 'int'>\n",
            "\n",
            "gc_content: 0.42\n",
            "type(gc_content): <class 'float'>\n",
            "\n"
          ]
        }
      ],
      "source": [
        "# Image you have a gene and you want to store various properties of this gene\n",
        "# such as Sample name, DNA length, GC content\n",
        "\n",
        "sample_name = \"E_coli_geneA\"\n",
        "dna_length = 5000\n",
        "gc_content = 0.421789\n",
        "\n",
        "print(f\"sample_name: {sample_name}\")\n",
        "print(f\"type(sample_name): {type(sample_name)}\\n\")\n",
        "\n",
        "print(f\"dna_length: {dna_length}\")\n",
        "print(f\"type(dna_length): {type(dna_length)}\\n\")\n",
        "\n",
        "print(f\"gc_content: {gc_content:.2f}\")\n",
        "print(f\"type(gc_content): {type(gc_content)}\\n\")\n"
      ]
    },
    {
      "cell_type": "markdown",
      "source": [
        "# Zooming into __int__ and __float__ datatypes"
      ],
      "metadata": {
        "id": "VSNEPHI4XJ5d"
      }
    },
    {
      "cell_type": "code",
      "source": [
        "# Biological Setup\n",
        "\n",
        "gene1_name = \"BRCA1\"\n",
        "gene2_name = \"TP53\"\n",
        "\n",
        "gene1_length = 1_863   # in base pairs\n",
        "gene2_length = 1_179\n",
        "\n",
        "gene1_counts = 25_430  # number of reads mapped\n",
        "gene2_counts = 21_000\n",
        "\n",
        "total_reads = 980_000\n",
        "\n",
        "# Challenge\n",
        "# a. Total gene length\n",
        "# b. Difference in gene length\n",
        "# c. calculate RPKM (Reads Per Kilobase of transcript per Million mapped reads)"
      ],
      "metadata": {
        "id": "SOzU9kJfXJWV"
      },
      "execution_count": 9,
      "outputs": []
    },
    {
      "cell_type": "code",
      "source": [
        "total_gene_length = gene1_length + gene2_length\n",
        "print(f\"total_gene_length: {total_gene_length}\")\n",
        "\n",
        "difference_in_gene_length = gene1_length - gene2_length\n",
        "print(f\"difference_in_gene_length: {difference_in_gene_length}\")\n",
        "\n",
        "gene1_rpkm = (gene1_counts * 1e9) / (gene1_length*total_reads)\n",
        "gene2_rpkm = (gene2_counts * 1e9) / (gene2_length*total_reads)\n",
        "\n",
        "print(f\"gene1_rpkm: {gene1_rpkm:.2f}\")\n",
        "print(f\"gene2_rpkm: {gene2_rpkm:.2f}\")"
      ],
      "metadata": {
        "colab": {
          "base_uri": "https://localhost:8080/"
        },
        "id": "isEdbU-KWZHa",
        "outputId": "28efd209-298f-487e-fd4a-e5b66af12202"
      },
      "execution_count": 13,
      "outputs": [
        {
          "output_type": "stream",
          "name": "stdout",
          "text": [
            "total_gene_length: 3042\n",
            "difference_in_gene_length: 684\n",
            "gene1_rpkm: 13928.60\n",
            "gene2_rpkm: 18175.21\n"
          ]
        }
      ]
    },
    {
      "cell_type": "markdown",
      "source": [
        "# Zooming into __string__ datatype"
      ],
      "metadata": {
        "id": "rS1rx62RhgnR"
      }
    },
    {
      "cell_type": "code",
      "source": [
        "# Biological Set up\n",
        "\n",
        "dna_seq1 = \"ATGCGATACGTTTATATATAGGCGCGCGCGCATATATAATTGCATGCTATAGCGCGGTATATATATATACGCGCGCGCTATTATTAT\"\n",
        "dna_seq2 = \"CGTATGCTATGCGCTAGCGGCTAGTGTCGCCTCTCACATGACGTAGCAGTGCACTCTCTCTCATCTACTACACGGTNCGCGCGCTAGC\"\n",
        "\n",
        "# Challenge\n",
        "# a. Calculate length of both dna\n",
        "# b. Concat the two DNA and print the resultant DNA\n",
        "# c. Print the resultant DNA in lower case\n",
        "# c. Print a chunck of the two DNA from nucleotide position 9 to 12\n",
        "# d. Calculate the GC contents of the resultant DNA\n",
        "# e. Find which DNA has Sequencing errors\n",
        "# f. Find Reverse complement of the two sequence"
      ],
      "metadata": {
        "id": "YsncJCqwfE18"
      },
      "execution_count": 15,
      "outputs": []
    },
    {
      "cell_type": "code",
      "source": [
        "print(f\"Length of dna_seq1: {len(dna_seq1)}\")\n",
        "print(f\"Length of dna_seq2: {len(dna_seq2)}\\n\")\n",
        "\n",
        "total_DNA = dna_seq1 + dna_seq2\n",
        "print(f\"total_DNA: {total_DNA}\\n\")\n",
        "\n",
        "print(f\"total_DNA in lower case: {total_DNA.lower()}\\n\")\n",
        "\n",
        "chunk_of_dna_seq1 = dna_seq1[8:12]\n",
        "chunk_of_dna_seq2 = dna_seq2[8:12]\n",
        "print(f\"chunk_of_dna_seq1: {chunk_of_dna_seq1}\")\n",
        "print(f\"chunk_of_dna_seq2: {chunk_of_dna_seq2}\\n\")\n",
        "\n",
        "len_total_DNA = len(total_DNA)\n",
        "gc_content = (total_DNA.count(\"G\") + total_DNA.count(\"C\")) / len_total_DNA\n",
        "print(f\"gc_content: {gc_content:.2f}\\n\")\n",
        "\n",
        "error_seq1 = dna_seq1.find(\"N\")\n",
        "error_seq2 = dna_seq2.find(\"N\")\n",
        "\n",
        "print(f\"error_seq1: {error_seq1}\")\n",
        "print(f\"error_seq2: {error_seq2}\\n\")\n",
        "\n",
        "complement_total_DNA = total_DNA.replace(\"A\", \"t\").replace(\"T\", \"a\").replace(\"G\", \"c\").replace(\"C\", \"g\").upper()[::-1]\n",
        "print(f\"complement_total_DNA: {complement_total_DNA}\")"
      ],
      "metadata": {
        "colab": {
          "base_uri": "https://localhost:8080/"
        },
        "id": "SxnGcJxMiO8U",
        "outputId": "a3517358-d6f9-471b-f150-b94217472a98"
      },
      "execution_count": 23,
      "outputs": [
        {
          "output_type": "stream",
          "name": "stdout",
          "text": [
            "Length of dna_seq1: 87\n",
            "Length of dna_seq2: 88\n",
            "\n",
            "total_DNA: ATGCGATACGTTTATATATAGGCGCGCGCGCATATATAATTGCATGCTATAGCGCGGTATATATATATACGCGCGCGCTATTATTATCGTATGCTATGCGCTAGCGGCTAGTGTCGCCTCTCACATGACGTAGCAGTGCACTCTCTCTCATCTACTACACGGTNCGCGCGCTAGC\n",
            "\n",
            "total_DNA in lower case: atgcgatacgtttatatataggcgcgcgcgcatatataattgcatgctatagcgcggtatatatatatacgcgcgcgctattattatcgtatgctatgcgctagcggctagtgtcgcctctcacatgacgtagcagtgcactctctctcatctactacacggtncgcgcgctagc\n",
            "\n",
            "chunk_of_dna_seq1: CGTT\n",
            "chunk_of_dna_seq2: ATGC\n",
            "\n",
            "gc_content: 0.49\n",
            "\n",
            "error_seq1: -1\n",
            "error_seq2: 76\n",
            "\n",
            "complement_total_DNA: GCTAGCGCGCGNACCGTGTAGTAGATGAGAGAGAGTGCACTGCTACGTCATGTGAGAGGCGACACTAGCCGCTAGCGCATAGCATACGATAATAATAGCGCGCGCGTATATATATATACCGCGCTATAGCATGCAATTATATATGCGCGCGCGCCTATATATAAACGTATCGCAT\n"
          ]
        }
      ]
    },
    {
      "cell_type": "markdown",
      "source": [
        "# Introducing Lists\n",
        "- Use lists to represent multiple genes, sequences, or measurements."
      ],
      "metadata": {
        "id": "LZljtG3QwFG2"
      }
    },
    {
      "cell_type": "code",
      "source": [
        "# Biological Set up\n",
        "\n",
        "genes = [\"myc\", \"nfkbiz\", \"kras\", \"brca1\", \"tp53\"]\n",
        "gc_values = [0.41, 0.56, 0.48, 0.67, 0.34]\n",
        "\n",
        "\n",
        "# Challenge\n",
        "# a. Calculate the number of genes we have in our list\n",
        "# b. Find the third gene in the list\n",
        "# c. Find the genes from position 2 to 4\n",
        "# d. Calculate the max gc_value and the gene corresponding to it\n",
        "# e. Calculate the min gc_value and the gene corresponding to it\n",
        "# f. calculate the average gc_value\n",
        "# g. Sort the gc_value in descending order and genes accordingly."
      ],
      "metadata": {
        "id": "qtkLSIPakhnH"
      },
      "execution_count": 27,
      "outputs": []
    },
    {
      "cell_type": "code",
      "source": [
        "print(f\"Number of genes in our list: {len(genes)}\\n\")\n",
        "print(f\"third gene in the list: {genes[2]}\\n\")\n",
        "print(f\"genes from position 2 to 4: {genes[1:4]}\\n\")\n",
        "\n",
        "max_gc_value = max(gc_values)\n",
        "max_gc_value_index = gc_values.index(max_gc_value)\n",
        "max_gc_value_gene = genes[max_gc_value_index]\n",
        "\n",
        "print(f\"max gc_value: {max_gc_value}\")\n",
        "print(f\"max gc_value gene: {max_gc_value_gene}\\n\")\n",
        "\n",
        "min_gc_value = min(gc_values)\n",
        "min_gc_value_index = gc_values.index(min_gc_value)\n",
        "min_gc_value_gene = genes[min_gc_value_index]\n",
        "\n",
        "print(f\"min gc_value: {min_gc_value}\")\n",
        "print(f\"min gc_value gene: {min_gc_value_gene}\\n\")\n",
        "\n",
        "print(f\"average gc_value: {sum(gc_values)/len(gc_values): .2f}\\n\")\n",
        "\n",
        "sorted_pairs = sorted(zip(gc_values, genes), reverse=True)\n",
        "sorted_gc_values, sorted_genes = zip(*sorted_pairs)\n",
        "print(f\"sorted_gc_values: {list(sorted_gc_values)}\")\n",
        "print(f\"sorted_genes: {list(sorted_genes)}\")"
      ],
      "metadata": {
        "colab": {
          "base_uri": "https://localhost:8080/"
        },
        "id": "1KpgEkcqxfp-",
        "outputId": "503c10a7-42b3-422f-9083-a72c2bae621f"
      },
      "execution_count": 30,
      "outputs": [
        {
          "output_type": "stream",
          "name": "stdout",
          "text": [
            "Number of genes in our list: 5\n",
            "\n",
            "third gene in the list: kras\n",
            "\n",
            "genes from position 2 to 4: ['nfkbiz', 'kras', 'brca1']\n",
            "\n",
            "max gc_value: 0.67\n",
            "max gc_value gene: brca1\n",
            "\n",
            "min gc_value: 0.34\n",
            "min gc_value gene: tp53\n",
            "\n",
            "average gc_value:  0.49\n",
            "\n",
            "sorted_gc_values: [0.67, 0.56, 0.48, 0.41, 0.34]\n",
            "sorted_genes: ['brca1', 'nfkbiz', 'kras', 'myc', 'tp53']\n"
          ]
        }
      ]
    },
    {
      "cell_type": "markdown",
      "source": [
        "# Introducing Dictionaries"
      ],
      "metadata": {
        "id": "CnGjV-CF35Z_"
      }
    },
    {
      "cell_type": "code",
      "source": [
        "# Biological Set up\n",
        "\n",
        "gene_data = {'genes':[\"myc\", \"nfkbiz\", \"kras\", \"brca1\", \"tp53\"],\n",
        "             'gc_values':[0.41, 0.56, 0.48, 0.67, 0.34]}\n",
        "\n",
        "extra_data = {'expression': [100, 200, 150, 50, 300]}\n",
        "\n",
        "# Challenge\n",
        "# a. Find all the keys of this dictionary\n",
        "# b. Find all the values of this dictionary\n",
        "# c. Find values corresponding to the keys\n",
        "# d. merge the gene_data with extra_data\n",
        "# e. Create a new dictionary mapping gene -> GC values\n",
        "# f. sort this new dictionary\n"
      ],
      "metadata": {
        "id": "n_qRMJ0H11d2"
      },
      "execution_count": 37,
      "outputs": []
    },
    {
      "cell_type": "code",
      "source": [
        "print(f\"gene_data keys: {gene_data.keys()}\")\n",
        "print(f\"gene_data values: {gene_data.values()}\\n\")\n",
        "\n",
        "print(f\"gene_data['genes']: {gene_data['genes']}\")\n",
        "print(f\"gene_data['gc_values']: {gene_data['gc_values']}\\n\")\n",
        "\n",
        "gene_data_copy = gene_data.copy()\n",
        "gene_data_copy.update(extra_data)\n",
        "print(f\"gene_data_copy: {gene_data_copy}\\n\")\n",
        "\n",
        "new_dict_genes = gene_data['genes']\n",
        "new_dict_gc_values = gene_data['gc_values']\n",
        "gene_gc_dict = dict(zip(new_dict_genes, new_dict_gc_values))\n",
        "print(f\"gene_gc_dict: {gene_gc_dict}\")\n",
        "\n",
        "sorted_gene_gc_dict = dict(sorted(gene_gc_dict.items(), key=lambda x: x[1], reverse=True))\n",
        "print(f\"sorted_gene_gc_dict: {sorted_gene_gc_dict}\")"
      ],
      "metadata": {
        "colab": {
          "base_uri": "https://localhost:8080/"
        },
        "id": "cHs7K_1c7YQt",
        "outputId": "e0e85406-bbc9-4c53-c0d6-9cb450e13c4e"
      },
      "execution_count": 41,
      "outputs": [
        {
          "output_type": "stream",
          "name": "stdout",
          "text": [
            "gene_data keys: dict_keys(['genes', 'gc_values'])\n",
            "gene_data values: dict_values([['myc', 'nfkbiz', 'kras', 'brca1', 'tp53'], [0.41, 0.56, 0.48, 0.67, 0.34]])\n",
            "\n",
            "gene_data['genes']: ['myc', 'nfkbiz', 'kras', 'brca1', 'tp53']\n",
            "gene_data['gc_values']: [0.41, 0.56, 0.48, 0.67, 0.34]\n",
            "\n",
            "gene_data_copy: {'genes': ['myc', 'nfkbiz', 'kras', 'brca1', 'tp53'], 'gc_values': [0.41, 0.56, 0.48, 0.67, 0.34], 'expression': [100, 200, 150, 50, 300]}\n",
            "\n",
            "gene_gc_dict: {'myc': 0.41, 'nfkbiz': 0.56, 'kras': 0.48, 'brca1': 0.67, 'tp53': 0.34}\n",
            "sorted_gene_gc_dict: {'brca1': 0.67, 'nfkbiz': 0.56, 'kras': 0.48, 'myc': 0.41, 'tp53': 0.34}\n"
          ]
        }
      ]
    },
    {
      "cell_type": "markdown",
      "source": [
        "# Introducing Loops and Decision Making"
      ],
      "metadata": {
        "id": "Sb4xB-08Amp6"
      }
    },
    {
      "cell_type": "code",
      "source": [
        "# Biological Set up\n",
        "\n",
        "sequences = [\"ATGCGC\", \"ATTTAA\", \"GCGCGC\", \"ATGCAT\"]\n",
        "\n",
        "# Challenge\n",
        "# a. Find the GC count for each DNA fragment in \"sequences\"\n",
        "# b. print \"High GC\" if GC count > 0.45 else print \"Low GC\""
      ],
      "metadata": {
        "id": "WBmKWwfO7c7C"
      },
      "execution_count": 43,
      "outputs": []
    },
    {
      "cell_type": "code",
      "source": [
        "threshold = 0.45\n",
        "for seq in sequences:\n",
        "    gc = (seq.count(\"G\") + seq.count(\"C\")) / len(seq)\n",
        "    if gc > threshold:\n",
        "        print(seq, \"→ High GC\")\n",
        "    else:\n",
        "        print(seq, \"→ Low GC\")"
      ],
      "metadata": {
        "colab": {
          "base_uri": "https://localhost:8080/"
        },
        "id": "MCIkXwm5BSIu",
        "outputId": "46961544-74bf-4832-a21b-6eb656922d30"
      },
      "execution_count": 44,
      "outputs": [
        {
          "output_type": "stream",
          "name": "stdout",
          "text": [
            "ATGCGC → High GC\n",
            "ATTTAA → Low GC\n",
            "GCGCGC → High GC\n",
            "ATGCAT → Low GC\n"
          ]
        }
      ]
    },
    {
      "cell_type": "code",
      "source": [],
      "metadata": {
        "id": "WMZGVcYBB4Cr"
      },
      "execution_count": null,
      "outputs": []
    }
  ]
}